File: apps/web/src/lib/pagination.ts
export const PAGE_SIZES = [10, 20, 50] as const;

export type PageSize = (typeof PAGE_SIZES)[number];

export type ResourcePage<T> = {
  items: T[];
  page: number | null;
  page_size: number;
  total: number;
  total_pages: number;
};

export function normalizeResourcePage<T>(
  value:
    | ResourcePage<T>
    | T[]
    | {
        items: T[];
        page?: number | null;
        page_size?: number;
        total?: number;
        total_pages?: number;
      },
  page: number,
  pageSize: PageSize,
): ResourcePage<T> {
  if (Array.isArray(value)) {
    return {
      items: value,
      page,
      page_size: pageSize,
      total: value.length,
      total_pages: value.length ? 1 : 0,
    };
  }
  return {
    items: value.items,
    page: value.page ?? page,
    page_size: value.page_size ?? pageSize,
    total: value.total ?? value.items.length,
    total_pages:
      value.total_pages ??
      (value.items.length
        ? Math.ceil((value.total ?? value.items.length) / pageSize)
        : 0),
  };
}

export function isPageSize(value: number): value is PageSize {
  return PAGE_SIZES.includes(value as PageSize);
}

export function visiblePages(page: number, totalPages: number) {
  if (totalPages <= 5) {
    return Array.from({ length: totalPages }, (_, index) => index + 1);
  }
  const start = Math.min(Math.max(page - 2, 1), totalPages - 4);
  return Array.from({ length: 5 }, (_, index) => start + index);
}

File: apps/control-api/src/agenttest/shared/pagination.py
from __future__ import annotations

from collections.abc import Sequence
from dataclasses import dataclass
from math import ceil

SUPPORTED_PAGE_SIZES = frozenset({10, 20, 50})


@dataclass(frozen=True, slots=True)
class PageRequest:
    page: int
    page_size: int

    def __post_init__(self) -> None:
        if self.page < 1:
            raise ValueError("page must be greater than or equal to 1")
        if self.page_size not in SUPPORTED_PAGE_SIZES:
            raise ValueError("page_size must be one of 10, 20, or 50")

    @property
    def offset(self) -> int:
        return (self.page - 1) * self.page_size


@dataclass(frozen=True, slots=True)
class PageResult[T]:
    items: list[T]
    total: int
    page: int
    page_size: int

    def __post_init__(self) -> None:
        if self.total < 0:
            raise ValueError("total must be greater than or equal to 0")
        PageRequest(page=self.page, page_size=self.page_size)

    @property
    def total_pages(self) -> int:
        return ceil(self.total / self.page_size) if self.total else 0


def paginate_items[T](items: Sequence[T], page_request: PageRequest) -> PageResult[T]:
    start = page_request.offset
    return PageResult(
        items=list(items[start : start + page_request.page_size]),
        total=len(items),
        page=page_request.page,
        page_size=page_request.page_size,
    )
